pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go

```diff
@@ -61,16 +61,20 @@ func (e *kubeEndpoints) Run(stopCh <-chan struct{}) {
 
 // processEndpointEvent triggers the config update.
 func processEndpointEvent(c *Controller, epc kubeEndpointsController, name string, namespace string, event model.Event, ep interface{}) error {
+	// Update internal endpoint cache no matter what kind of service, even headless service.
+	// As for gateways, the cluster discovery type is `EDS` for headless service.
+	updateEDS(c, epc, ep, event)
```

Comment:
nit: I think we can just call the EDSCacheUpdate for headless and since we are doing full push it would take care of pushing endpoints - might require more refactoring though

Reply:
This is also my thought, want to refactor the endpointShard update logic. Currently, for k8s registry, we only update from the endpoint/svc handler, i wanna call all update shards during push along with push context init. And we have recorded the svcs updated, so we only need to do a partial updates there.

By this we can save many cpu cycles, especially when endpoints update frequently.

Reply:
Makes sense. Let us do that in a follow-up